core/src/main/java/org/eclipse/hono/util/MessageHelper.java

```diff
@@ -527,4 +557,44 @@ public static void setCreationTime(final Message msg) {
         }
     }
 
+    /**
+     * Construct a message consumer for event messages that detects an indication that a device is ready to receive an upstream
+     * message and delegates to the passed callback handlers.
+     * <p>
+     * Note that there are two different ways how this can be detected:
+     * <ul>
+     *     <li>a specific event (that has the Content-Type {@link EventConstants#CONTENT_TYPE_EMPTY_NOTIFICATION}</li>
+     *     <li>any other event that carries the AMQP 1.0 application property {@link MessageHelper#APP_PROPERTY_DEVICE_TTD}</li>
+     * </ul>
+     *
+     * @param tenantId The tenant this handler is scoped to.
+     * @param messageConsumer Message consumer that handles generic messages.
+     * @param notificationReadyToDeliverConsumer BiConsumer&lt;String,String&gt; that is called when the message indicates the current readiness
+     *                                           to receive an upstream message. Maybe {@code null}. Provided is the tenantId
+     *                                           and deviceId.
+     * @return The message consumer that invokes the specific handlers depending on the event type.
+     */
+    public static Consumer<Message> createMessageDemuxHandler(final String tenantId,
+                                                                    final Consumer<Message> messageConsumer,
+                                                                    final Consumer<NotificationDeviceReadyToDeliver> notificationReadyToDeliverConsumer
+    ) {
+
+
+        return (msg -> {
+            final Section body = msg.getBody();
+            if (!(body instanceof Data)) {
+                return;
+            }
+
+            Optional.ofNullable(notificationReadyToDeliverConsumer).map(notificationConsumer -> {
+                NotificationDeviceCommandReadyConstants.isDeviceCurrentlyReadyForCommands(tenantId, msg).ifPresent(
+                        notificationObject -> notificationConsumer.accept(notificationObject));
+
+                messageConsumer.accept(msg);
+
+                return notificationConsumer;
+            }).orElse(v -> messageConsumer.accept(msg));
```

Comment:
what's the purpose of passing in `null` as the notificationConsumer? Wouldn't the client in that case simply register a Consumer<Message> using HonoClient?

Reply:
Good question - I thought of extending this to more callbacks when we merge this with the "connected/disconnected" notification. And then it might be good to only provide the callbacks that an application is interested in.

But thinking about it now it seems to make things more complicated and others will ask the same question as you. So it probably is better to enforce callbacks to be always not null, and in some cases to have empty implementations instead.
Will remove the "maybe null" from the code.